### PR TITLE
drivers: modem: gsm_ppp: Add RTC query

### PR DESCRIPTION
**Pull request by ycsin**
Added RTC query using AT+CCLK command to the gsm_ppp driver, see `gsm_ppp_get_local_time`

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

### CODE REVIEW

**Comment by ycsin** [comment]
The functions are basically copied from `hl7800` driver by @rerickson1, I have added the copyright to the header. Tell me if I need to add anything else.

Would be great if we can somehow have a shared file for all the modem drivers for generic AT commands since they will be the same across all modems.

**Review by jukkar — CHANGES_REQUESTED** [review]
If the RTC query functionality can be shared between different modems, then it should be made generic enough so that different modems can share the code.

Bigger issue here is that the time functionality should be integrated to APIs that return wall clock time instead of adding a time get function to modem which does not really care about time/date information.

Currently there is no public driver API for modems so we would need to create one, and then perhaps integrate this RTC (time/date) and GNSS (location) functionality to that API. Then the wall clock time API functions could use the modem driver API to fetch the current time in a generic way.

**Comment by ycsin** [comment]
@jukkar 
>If the RTC query functionality can be shared between different modems, then it should be made generic enough so that different modems can share the code.

The RTC query functionality here should just work for any modem driver that used delimiter to parse the AT commands, that includes at least BG9x. In fact the code here is copied from my offline bg9x driver, which is again copied and modified from the hl7800 driver.

**Comment by ycsin** [comment]
Will fix the buildkite error probably tomorrow

**Comment by jukkar** [comment]
After thinking this more, imho the time functions should be hidden behind a driver (like is done for ptp clock for example). Then GSM modem driver could implement this driver and provide time functionality to the callers. A generic API would then just query the driver found in the system in order to get time information.

**Comment by ycsin** [comment]
Would it be possible to have a generic AT driver for all the common AT commands as a base that works for either uart / uart_mux where everything can be built on top, such as the setup of a modem, ppp, offloaded sockets, query of RSSI, time query functions/API, GNSS APIs.

I just think that a lot of things could be shared among modems such as this pr (converted from hl7800, applicable to at least bg9x driver and gsm_ppp), querying the rssi/registration periodically, handling of no-signal scenario, the recovery from a non-responding hardware, sms, etc. Having a shared AT base can potentially reduce quite a bit of work.

But I can't imagine how things would look like and if this is actually feasible.

**Comment by ycsin** [comment]
Please remove the stale label

**Comment by carlescufi** [comment]
> I wonder if its possible to have another file that can access the UART of the modem and work alongside with the modem so that we can do all the common AT command stuff there like RTC/RSSI/network type/SMS/etc and can be shared by all modem?

I think that it's time we try to standardize some of these modem functionality into a common API, yes. @rerickson1 do you know if you could send a proposal?

**Comment by rerickson1** [comment]
> > I wonder if its possible to have another file that can access the UART of the modem and work alongside with the modem so that we can do all the common AT command stuff there like RTC/RSSI/network type/SMS/etc and can be shared by all modem?
> 
> I think that it's time we try to standardize some of these modem functionality into a common API, yes. @rerickson1 do you know if you could send a proposal?

`drivers/modem/modem_cmd_handler.c`, `drivers/modem/modem_context.c`, `drivers/modem/modem_iface_uart.c`, `drivers/modem/modem_pin.c`, `drivers/modem/modem_socket.c` are all the **start of a common modem framework**. The quectel-bg9x and sara-r4 drivers use this newer framework. The HL7800 and wncm14a2a drivers do not use this new framework yet.

**Comment by ycsin** [comment]
One small heads-up for `gsm_ppp` is that it supports more than one brand and certainly quite a few models so while all of these modem supports ppp more or less the same (slightly different commands to enter ppp mode), the way they do FOTA/GNSS etc can have some significant difference.

I'm pondering on the possibility of having a generic "at_core" that can use an (uart) interface and does generic AT commands which can be used like a "plugin" for other offload/ppp modem drivers that exposes the (uart) interface. And same goes for the `gsm_ppp`, which should be generic enough and other brand (SIMCOM/QUECTEL) should just be a "plugin" of it.

**Review by mniestroj — CHANGES_REQUESTED** [review]

File: drivers/modem/gsm_ppp.c [hunk 1]
```diff
@@ -22,6 +24,10 @@ LOG_MODULE_REGISTER(modem_gsm, CONFIG_MODEM_LOG_LEVEL);
 #include "modem_cmd_handler.h"
 #include "../console/gsm_mux.h"
 
+#ifdef CONFIG_NEWLIB_LIBC
+#include <time.h>
+#endif
+
```
[the remark this answers is not in the document]
Comment (rerickson1) [reply drivers/modem/gsm_ppp.c:30]: `time.h` gives access to the `struct tm`. This gives us a standard way to represent time. Maybe there is a better option?

[the remark this answers is not in the document]
Comment (mniestroj) [reply drivers/modem/gsm_ppp.c:30]: We do have `zephyr/lib/libc/minimal/include/time.h` with `struct tm` for minimal libc, so why do we enable this part for newlib libc only?

[the remark this answers is not in the document]
Comment (rerickson1) [reply drivers/modem/gsm_ppp.c:30]: agree, I don't see a reason. When we added this to `hl7800.c` there may have been a better reason, but that was a long while ago.

[the remark this answers is not in the document]
Comment (rerickson1) [reply drivers/modem/gsm_ppp.c:30]: https://github.com/zephyrproject-rtos/zephyr/pull/40137

File: drivers/modem/Kconfig.gsm [hunk 2]
```diff
@@ -34,6 +34,13 @@ config MODEM_GSM_QUECTEL
 
 endchoice
 
+config MODEM_GSM_PTP_CLOCK
```
Comment (jukkar) [inline drivers/modem/Kconfig.gsm:37]: This looks better now but actually the GSM modem most likely cannot really provide PTP (Precision Time Protocol) support which is typically only implemented for Ethernet. I was suggesting earlier that something similar like what ptp_clock driver is implemented, not exactly that we directly use ptp_clock. For example for modem clock driver, there is no need to have support for adjust() or rate_adjust(), also probably the set() is also not needed (we most probably cannot set the time in the modem, just read what the current time is).

File: drivers/modem/gsm_ppp.c [hunk 3]
```diff
@@ -1,5 +1,7 @@
 /*
  * Copyright (c) 2020 Intel Corporation
+ * Copyright (c) 2020 Laird Connectivity
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c:3]: Why are you adding this Laird Connectivity copyright from 2020 here?

Comment (ycsin) [reply drivers/modem/gsm_ppp.c:3]: @ycsin Basically the timestring conversion is copied from the hl7800 driver, with some modifications

File: drivers/modem/Kconfig.gsm (hunk 2 above)
Comment (ycsin) [reply drivers/modem/Kconfig.gsm:37]: > the GSM modem most likely cannot really provide PTP

Yeah, I'm literally hijacking the ptp driver API here. Not really sure what is the best path forward.

>Currently there is no public driver API for modems so we would need to create one, and then perhaps integrate this RTC (time/date) and GNSS (location) functionality to that API. Then the wall clock time API functions could use the modem driver API to fetch the current time in a generic way.

I recently worked with the esp8266 wifi driver and needed the same function, maybe we need an AT driver API, but I'm not sure

**Pull request closed without merging**